src/computation.ts

```diff
@@ -98,125 +116,147 @@ export class Computation {
     this.run();
   }
 
-  private run(): void {
-    // Guard against concurrent processing
-    if (this.isProcessing) {
-      return;
-    }
-
-    this.isProcessing = true;
-
-    if (this.eventQueue.length === 0) {
+  private completeTask(result: CompResult) {
+    util.assertDefined(this.task);
+    this.network.send({ kind: "completeTask", id: this.task.id, counter: this.task.counter }, () => {
+      // Once the task is completed reset the computation
+      this.task = undefined;
+      this.eventQueue = [];
       this.isProcessing = false;
-      return;
-    }
-
-    const event = this.eventQueue.shift();
-    util.assertDefined(event);
-
-    const { fn, args } = this.invocationParams!;
-
-    Coroutine.exec(this.promiseId, fn, args, this.handler, (result) => {
-      if (result.type === "completed") {
-        this.handler.resolvePromise(this.promiseId, result.value, (durablePromise) => {
-          util.assertDefined(this.task);
-          this.completeTask({
-            kind: "completed",
-            promiseId: this.promiseId,
-            result: durablePromise.value,
-          });
-        });
-      } else {
-        if (result.localTodos.length === 0) {
-          this.handleRemoteTodos(this.promiseId, result.remoteTodos);
-        } else {
-          this.handleLocalTodos(result.localTodos);
-        }
-      }
+      this.seenTodos.clear();
+      this.callback?.(result);
+      this.callback = undefined;
     });
-
-    // Reset processing flag and continue if there are more events
-    this.isProcessing = false;
-    if (this.eventQueue.length > 0) {
-      this.run();
-    }
   }
 
-  private handleRemoteTodos(rootId: string, remoteTodos: RemoteTodo[]): void {
+  private handleRemoteTodos(todos: RemoteTodo[]) {
     let createdCallbacks = 0;
-    const totalCallbacks = remoteTodos.length;
+    const totalCallbacks = todos.length;
+    // reset the event queue, if we made it here it means that all localTodos were proccesed,
+    // if there is anything else in the queue it is redundant work and we should clean it to
+    // avoid races or problems due to concurrency
+    this.eventQueue = [];
```

Comment:
If there is an entry in the event queue that means we have already received a resume message. But we can safely clear that event because we will get a completed promise when attempting to create a callback which will effectively replace the resume message.

Is this accurate? If so I understand how this avoids redundancy, but I don't see how this avoids a race conditions - the event queue itself guards against this and it would be safe to call run again?